### PR TITLE
Configure context to prevent a hanging call

### LINKED ISSUE #10506: cilium-agent hangs on startup and provides no useful logging
[cilium-sysdump-20200306-215302.zip](https://github.com/cilium/cilium/files/4300663/cilium-sysdump-20200306-215302.zip)

## Bug report

**General Information**

- Cilium version 1.16.3
- Kernel version 5.3.18
- Kubernetes 1.16.2
- Deployed quick-install manifest https://raw.githubusercontent.com/cilium/cilium/v1.6.3/install/kubernetes/quick-install.yaml

**How to reproduce the issue**

1. Deploy Kubernetes 1.16.2
2. Run 'kubectl create -f https://raw.githubusercontent.com/cilium/cilium/v1.6.3/install/kubernetes/quick-install.yaml'
3. Cilium agent starts and registers healthy on the master node. However, on worker nodes it starts but never registers as healthy.

`....
level=info msg="Initializing daemon" subsys=daemon
level=info msg="Detected MTU 9000" subsys=mtu
level=debug msg="creating new EventQueue" name=repository-change-queue numBufferedEvents=100 subsys=eventqueue
level=debug msg="creating new EventQueue" name=repository-reaction-queue numBufferedEvents=100 subsys=eventqueue
level=debug msg="Performing regular background work" subsys=nodemanager syncInterval=1m0s
level=debug msg="Registered BPF map" path=/run/cilium/bpffs/tc/globals/cilium_lxc subsys=bpf
level=debug msg="Registered BPF map" path=/run/cilium/bpffs/tc/globals/cilium_ipcache subsys=bpf
level=debug msg="Registered BPF map" path=/run/cilium/bpffs/tc/globals/cilium_metrics subsys=bpf
level=debug msg="Registered BPF map" path=/run/cilium/bpffs/tc/globals/cilium_tunnel_map subsys=bpf
level=debug msg="Registered BPF map" path=/run/cilium/bpffs/tc/globals/cilium_lb4_services_v2 subsys=bpf
level=debug msg="Registered BPF map" path=/run/cilium/bpffs/tc/globals/cilium_lb4_backends subsys=bpf
level=debug msg="Registered BPF map" path=/run/cilium/bpffs/tc/globals/cilium_lb4_reverse_nat subsys=bpf
level=debug msg="Registered BPF map" path=/run/cilium/bpffs/tc/globals/cilium_lb4_rr_seq_v2 subsys=bpf
level=debug msg="Starting new controller" name=metricsmap-bpf-prom-sync subsys=controller uuid=c163df0a-5ff3-11ea-b2b4-0cc47a6f229c
level=debug msg="Restored backend IDs" backendIDs="map[]" subsys=daemon
level=debug msg="Controller func execution time: 84.275µs" name=metricsmap-bpf-prom-sync subsys=controller uuid=c163df0a-5ff3-11ea-b2b4-0cc47a6f229c
level=info msg="Restore service IDs from BPF maps" duration="211.036µs" failed=0 removed=0 restored=0 skipped=0 subsys=daemon
level=info msg="Removing stale endpoint interfaces" subsys=daemon
level=info msg="Establishing connection to apiserver" host="https://10.96.0.1:443" subsys=k8s
level=info msg="Connected to apiserver" subsys=k8s
level=debug msg="Controller func execution time: 245.316µs" name=metricsmap-bpf-prom-sync subsys=controller uuid=c163df0a-5ff3-11ea-b2b4-0cc47a6f229c
level=debug msg="Controller func execution time: 160.975µs" name=metricsmap-bpf-prom-sync subsys=controller uuid=c163df0a-5ff3-11ea-b2b4-0cc47a6f229c
level=debug msg="Controller func execution time: 171.002µs" name=metricsmap-bpf-prom-sync subsys=controller uuid=c163df0a-5ff3-11ea-b2b4-0cc47a6f229c
level=debug msg="Controller func execution time: 211.557µs" name=metricsmap-bpf-prom-sync subsys=controller uuid=c163df0a-5ff3-11ea-b2b4-0cc47a6f229c
level=debug msg="Controller func execution time: 207.11µs" name=metricsmap-bpf-prom-sync subsys=controller uuid=c163df0a-5ff3-11ea-b2b4-0cc47a6f229c
level=debug msg="Controller func execution time: 257.866µs" name=metricsmap-bpf-prom-sync subsys=controller uuid=c163df0a-5ff3-11ea-b2b4-0cc47a6f229c
level=debug msg="Controller func execution time: 260.612µs" name=metricsmap-bpf-prom-sync subsys=controller uuid=c163df0a-5ff3-11ea-b2b4-0cc47a6f229c
level=debug msg="Controller func execution time: 251.174µs" name=metricsmap-bpf-prom-sync subsys=controller uuid=c163df0a-5ff3-11ea-b2b4-0cc47a6f229c
level=debug msg="Controller func execution time: 281.478µs" name=metricsmap-bpf-prom-sync subsys=controller uuid=c163df0a-5ff3-11ea-b2b4-0cc47a6f229c
level=debug msg="Controller func execution time: 270.928µs" name=metricsmap-bpf-prom-sync subsys=controller uuid=c163df0a-5ff3-11ea-b2b4-0cc47a6f229`

There are few issues I see:

1. Most importantly, the cilium-agent never seems to come online despite no errors being registered in the logs.
2. I have enabled debug logging, and I don't get enough information from it to point me in a helpful direction. Admittedly, this could be that I don't know what I'm doing :smile: It could also mean that a little more instrumentation is needed. The logs give me no indication as to what cilium-agent is doing and what next steps might be.

### PR DESCRIPTION
**Pull request by manuelbuil**
When using interfaces with MTU=9000, we sometimes see how cilium-agent
gets stuck when doing this Rest call. Eventually it gets killed by
Kubernetes because it exceeds the liveness probe and as a result there
are no useful logs.

This patch uses contexts to provide a timeout to the call. The 120s are
used since this number is defined in one of the examples for the
livenessProbe.initialDelaySeconds

https://github.com/cilium/cilium/blob/master/install/kubernetes/quick-install.yaml#L364

We can discuss if perhaps makes sense to generalize all calls to include
such a timeout with a globally defined timeout variable

Fixes: #10506

Signed-off-by: Manuel Buil <mbuil@suse.com>

Please ensure your pull request adheres to the following guidelines:

- [X] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [X] All code is covered by unit and/or runtime tests where feasible.
- [X] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [X] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [X] Provide a title or release-note blurb suitable for the release notes.
- [X] Thanks for contributing!

Fixes: #10506

```release-note
<!-- Enter the release note text here if needed or remove this section! -->
```

### CODE REVIEW

File: pkg/k8s/node.go [hunk 1]
```diff
@@ -209,8 +217,25 @@ func ParseNode(k8sNode *types.Node, source source.Source) *node.Node {
 // GetNode returns the kubernetes nodeName's node information from the
 // kubernetes api server
 func GetNode(c kubernetes.Interface, nodeName string) (*v1.Node, error) {
-	// Try to retrieve node's cidr and addresses from k8s's configuration
-	return c.CoreV1().Nodes().Get(context.TODO(), nodeName, metav1.GetOptions{})
+	// Pass a context so that the call fails if it takes more than 120 seconds
+	ctx, cancel := context.WithTimeout(context.TODO(), 120*time.Second)
```
Comment (vadorovsky) [inline pkg/k8s/node.go:221]: https://github.com/cilium/cilium/blob/master/pkg/defaults/defaults.go is a good place to store timeout values

**Review by vadorovsky — CHANGES_REQUESTED** [review]
I would prefer to have a more generic solution. For example, using `context.WithTimeout` to all client calls (should be possible, since we already wrap the client in `pkg/k8s`). Or at least using `context.WithTimeout` in all calls which are done during daemon launch. I don't have a clear idea how to do that though, I will think about it.

File: pkg/k8s/node.go [hunk 2]
```diff
@@ -209,8 +217,25 @@ func ParseNode(k8sNode *types.Node, source source.Source) *node.Node {
...
+		result.nodeInfo, result.requestError = c.CoreV1().Nodes().Get(ctx, nodeName, metav1.GetOptions{})
+		channel <- *result
+		}()
+	select {
+	case <-ctx.Done():
+		scopedLog.WithError(err).WithField("c.CoreV1().Nodes().Get").Error("Deadline exceeded when fetching the nodeName's information")
+		return nil, ctx.Err()
+	case request := <-channel:
+		return request.nodeInfo, request.requestError
+        }
```
Comment (aanm) [inline pkg/k8s/node.go:238]: Why the use of channels? the context itself should be enough to timeout this request.

**Review by aanm — CHANGES_REQUESTED** [review]
>When using interfaces with MTU=9000, we sometimes see how cilium-agent
gets stuck when doing this Rest call.

Can you provide more context why is this happening?

**Comment by manuelbuil** [comment]
I can't reproduce the issue anymore. I suspect it got fixed with the client-go update. Therefore, I close this

**Pull request closed without merging**